modules/diskq/logqueue-disk-non-reliable.c

```diff
@@ -649,6 +651,139 @@ _push_tail(LogQueue *s, LogMessage *msg, const LogPathOptions *path_options)
+   * anyway, and any sane number decreased the performance measurably.
+   *
+   * This means that per-thread input queues contain _all_ items that
+   * a single poll iteration produces. And once the reader is finished
+   * (either because the input is depleted or because of
+   * log_fetch_limit / window_size) the whole bunch is shared between the
+   * front_cache and legacy behaviour.
+   */
+
+  if (thread_index >= 0)
+    {
+      /* fastpath, use per-thread input FIFOs */
+      if (!self->input_queues[thread_index].finish_cb_registered)
+        {
+          /* this is the first item in the input FIFO, register a finish
+           * callback to make sure it gets moved to the wait_queue if the
+           * input thread finishes
+           * One reference should be held, while the callback is registered
+           * avoiding use-after-free situation
+           */
+
+          main_loop_worker_register_batch_callback(&self->input_queues[thread_index].cb);
+          self->input_queues[thread_index].finish_cb_registered = TRUE;
+          log_queue_ref(&self->super.super);
+        }
+
+      log_msg_write_protect(msg);
+      node = log_msg_alloc_queue_node(msg, path_options);
+      iv_list_add_tail(&node->list, &self->input_queues[thread_index].items);
+
+      self->input_queues[thread_index].len++;
+
+      log_msg_unref(msg);
+      return;
+    }
+  // slow path
+  _push_tail_single_message(s, msg, path_options);
+}
+
+static void
+_move_part_of_input_to_front_cache(LogQueueDiskNonReliable *self, InputQueue *input_queue,
+                                   gint num_msgs_to_send_to_front_cache)
+{
+  struct iv_list_head *head = &input_queue->items;
+  struct iv_list_head *pos = head->next;
+  gint count = 0;
+  g_mutex_lock(&self->super.super.lock);
+  while (pos != head && count < num_msgs_to_send_to_front_cache)
+    {
+      struct iv_list_head *next = pos->next;
+      LogMessageQueueNode *node = iv_list_entry(pos, LogMessageQueueNode, list);
+      LogPathOptions lpo = LOG_PATH_OPTIONS_INIT;
+      lpo.ack_needed = node->ack_needed;
+      lpo.flow_control_requested = node->flow_control_requested;
+
+      LogMessage *msg = log_msg_ref(node->msg);
+      _push_tail_front_cache(self, msg, &lpo);
+      log_msg_free_queue_node(node);
+
+      iv_list_del_init(pos);
+      input_queue->len--;
+      log_queue_queued_messages_inc(&self->super.super);
+
+      pos = next;
+      count++;
+    }
+  log_queue_push_notify(&self->super.super);
+  g_mutex_unlock(&self->super.super.lock);
+}
+
+static void
+_move_part_of_input_to_disk_or_flow_control_window(LogQueueDiskNonReliable *self, InputQueue *input_queue)
+{
+  LogMessageQueueNode *node = iv_list_entry(input_queue->items.next, LogMessageQueueNode, list);
+  LogPathOptions lpo = LOG_PATH_OPTIONS_INIT;
+  lpo.ack_needed = node->ack_needed;
+  lpo.flow_control_requested = node->flow_control_requested;
+  LogMessage *msg = log_msg_ref(node->msg);
```

Comment:
`_extract_queue_node()`

similarly to the previous case, I don't think we need `flow_control_requested` propagation